modules/hpe_nimble_volume.py

```diff
@@ -0,0 +1,865 @@
+
+    module = AnsibleModule(argument_spec=fields, required_if=required_if)
+    if client is None:
+        module.fail_json(msg='the python nimble_sdk module is required')
+
+    state = module.params["state"]
+    vol_name = module.params["name"]
+    size = module.params["size"]
+    description = module.params["description"]
+    perfpolicy_name = module.params["perfpolicy_name"]
+    limit = module.params["limit"]
+    online = module.params["online"]
+    owned_by_group_name = module.params["owned_by_group_name"]
+    multi_initiator = module.params["multi_initiator"]
+    iscsi_target_scope = module.params["iscsi_target_scope"]
+    pool_name = module.params["pool_name"]
+    read_only = module.params["read_only"]
+    block_size = module.params["block_size"]
+    clone = module.params["clone"]
+    agent_type = module.params["agent_type"]
+    dest_pool_name = module.params["dest_pool_name"]
+    cache_pinned = module.params["cache_pinned"]
+    thinly_provisioned = module.params["thinly_provisioned"]
+    encryption_cipher = module.params["encryption_cipher"]
+    app_uuid = module.params["app_uuid"]
+    folder_name = module.params["folder_name"]
+    dedupe_enabled = module.params["dedupe_enabled"]
+    limit_iops = module.params["limit_iops"]
+    limit_mbps = module.params["limit_mbps"]
+    parent_vol_name = module.params["parent_vol_name"]  # used for cloning
+    snapshot_name = module.params["snapshot_name"]
+    volcoll_name = module.params["volcoll_name"]
+    metadata = module.params["metadata"]
+    force = module.params["force"]
+    force_vvol = module.params["force_vvol"]
+    move = module.params["move"]
+    hostname = module.params["hostname"]
+    username = module.params["username"]
+    password = module.params["password"]
+
+    if (username is None or password is None or hostname is None):
+        module.fail_json(msg="Volume creation failed. Storage system IP or username or password is null")
+
+    client_obj = client.NimOSClient(
+        hostname,
+        username,
+        password
+    )
+    # defaults
+    return_status = changed = False
+    msg = "No Task was run"
+
+    # States
+    if move is True and state == "present":
+        if utils.is_null_or_empty(dest_pool_name) is False:
+            return_status, changed, msg, changed_attrs_dict = move_volume(client_obj, vol_name, dest_pool_name, force_vvol)
+        else:
+            module.fail_json(msg="Volume Move failed. dest_pool_name is null")
+
+    elif state == "create" or state == "present":
+        # state create/present can be provided for creating a new volume or
+        # creating a clone from source volume
+        if parent_vol_name is not None:
+            return_status, changed, msg, changed_attrs_dict = clone_volume(
+                client_obj, parent_vol_name, state,
+                vol_name, snapshot_name)
+        else:
+            if utils.is_null_or_empty(vol_name):
```

Comment:
I think this empty 'vol_name' validation can be moved before # States so that its applicable for all cases.

Reply:
ya.. makes sense